File: sudoku.rs
// Rust sudoku implementation
// This is a heuristic guided search with backout like the Nim version

use clap::Parser;
use std::assert;
use std::fs;
use std::fs::read_to_string;
use std::sync::{Arc, Mutex};
use std::thread;
use std::thread::available_parallelism;

// TODO
//   Threading?
//   Some structure for iterating repetition, passing in a closure or using a macro?
//   Check if compile time index calculation is possible now.

// The field type is a bitfield where each of the bits represents the possibility that the
// corresponding digit could be found there.  For instance, if the bitfield is 0x03 then either a 1
// or a 2 could be there.
type Field = u16;

#[derive(Copy, Clone)]
struct Sudoku {
    // The final derived value for each cell. 0 if still unknown or a bitfield with a single
    // bit set if the solution has been found. Stored as a field to better interface with
    // other arrays and improve performance for set_value and clear_value.
    cells: [Field; 81],

    // Remaining possibilities for each block, column, or row.
    blocks: [Field; 9],
    cols: [Field; 9],
    rows: [Field; 9],

    // Number of remaining unsolved cells
    remaining: usize,
}

impl Sudoku {
    fn print(&self) {
        let mut s = "".to_string();
        for (i, v) in self.into_iter().enumerate() {
            if i == 0 {
            } else if i % 27 == 0 {
                s.push_str("\n\n");
            } else if i % 9 == 0 {
                s.push('\n');
            } else if i % 3 == 0 {
                s.push_str("   ");
            } else {
                s.push(' ');
            }
            s.push_str(&field_to_val(*v).to_string());
        }
        println!("{s}\n\n");
    }

    fn set_value(&mut self, val: Field, location: usize) {
        assert!(self.cells[location] == 0);
        self.cells[location] = val;

        let (blk, col, row) = get_indices(location);
        self.blocks[blk] &= !val;
        self.cols[col] &= !val;
        self.rows[row] &= !val;
        self.remaining -= 1;
    }

    fn clear_value(&mut self, location: usize) {
        assert!(self.cells[location] != 0);
        let val = self.cells[location];
        self.cells[location] = 0;

        let (blk, col, row) = get_indices(location);
        self.blocks[blk] |= val;
        self.cols[col] |= val;
        self.rows[row] |= val;
        self.remaining += 1;
    }

    fn back_out_moves(&mut self, eager_moves: Vec<usize>) {
        for guess in eager_moves {
            self.clear_value(guess);
        }
    }

    fn solve(&mut self) -> bool {
        // Stage once, do as many forced moves as possible
        // Keep trying until they stop coming
        let mut eager_moves: Vec<usize> = Vec::with_capacity(16);

        loop {
            let mut solved_count = 0;
            let mut lowest_space: usize = 0;
            let mut lowest_count: u32 = std::u32::MAX;
            let mut lowest_possibles: Field = 0;
            for i in 0..81 {
                if self.cells[i] != 0 {
                    // this cell is already solved
                    continue;
                }
                let (b, c, r) = get_indices(i);
                let possibles = self.blocks[b] & self.cols[c] & self.rows[r];
                let count: u32 = possibles.count_ones();

                if count == 1 {
                    self.set_value(possibles, i);
                    solved_count += 1;
                    eager_moves.push(i);
                } else if count == 0 {
                    // We're down a blind ally, abort
                    self.back_out_moves(eager_moves);
                    return false;
                } else if count < lowest_count {
                    lowest_count = count;
                    lowest_space = i;
                    lowest_possibles = possibles;
                }
            }
            if solved_count != 0 {
                // If we're finding moves by elimination don't start guessing yet, just keep on
                // solving this way
                continue;
            } else if self.remaining == 0 {
                // We won!
                return true;
            }
            let possibles = field_to_vals(lowest_possibles);
            for guess in possibles {
                self.set_value(val_to_field(guess), lowest_space);
                if self.solve() {
                    return true;
                }
                self.clear_value(lowest_space);
            }
            self.back_out_moves(eager_moves);
            // None of the guesses worked, we're on a bad branch.  Abort
            return false;
        }
    }
}

fn val_to_field(x: u8) -> Field {
    1 << (x - 1)
}

fn field_to_val(x: Field) -> u8 {
    if x == 0 {
        return 0;
    }
    assert!(x.count_ones() == 1);
    return x.trailing_zeros() as u8 + 1;
}

fn field_to_vals(x: Field) -> Vec<u8> {
    // In theory, the maximum number of ones returned from count_ones() on a 16 bit number should
    // be 16, which should fit in a u8, which would let me use `usize::from` here instead of `as`,
    // sadly it returns a u32 for some reason so we have to do something dangerous looking to the
    // compiler.
    let mut ret: Vec<u8> = Vec::with_capacity(x.count_ones() as usize);
    if x == 0 {
        return ret;
    }

    let mut x = x;
    let mut val = 1u8;

    while x > 0 {
        if 1 == (x % 2) {
            ret.push(val);
        }
        val += 1;
        x >>= 1;
    }
    ret
}

fn get_indices(i: usize) -> (usize, usize, usize) {
    let row = i / 9;
    let col = i % 9;
    let blk = row / 3 + 3 * (col / 3);
    (blk, col, row)
}

fn load_sudoku(file: &String) -> Sudoku {
    // Bits 1 through 9 are set since any of those might be a valid guess to start with.
    let start_possibilties: Field = 0x1FF;
    let mut s = Sudoku {
        cells: [0; 81],
        blocks: [start_possibilties; 9],
        cols: [start_possibilties; 9],
        rows: [start_possibilties; 9],
        remaining: 81,
    };

    let contents = fs::read_to_string(file).expect("Something went wrong reading the file");
    for (i, c) in (0..81).zip(contents.split_whitespace()) {
        let x = c.parse().unwrap();
        if x == 0 {
            continue;
        }
        s.set_value(val_to_field(x), i)
    }

    return s;
}

impl<'a> IntoIterator for &'a Sudoku {
    type Item = &'a Field;
    type IntoIter = std::slice::Iter<'a, Field>;

    fn into_iter(self) -> Self::IntoIter {
        self.cells.iter()
    }
}

fn solve_n(filename: &String, n: usize) -> Sudoku {
    // Solves a sudoku N times, and then returns the last Sudoku
    // Load the initial state once and copy it for each iteration
    let initial_sudoku = load_sudoku(&filename);
    let mut s = initial_sudoku.clone();
    s.solve();
    for _ in 1..n {
        s = initial_sudoku.clone();
        s.solve();
    }
    return s;
}

fn solve_alt_format(filename: &String) {
    let mut count = 0;
    for line in read_to_string(filename).unwrap().lines() {
        if line.len() != 81 {
            println!("Solving {line}");
            continue;
        }

        let start_possibilties: Field = 0x1FF;
        let mut s = Sudoku {
            cells: [0; 81],
            blocks: [start_possibilties; 9],
            cols: [start_possibilties; 9],
            rows: [start_possibilties; 9],
            remaining: 81,
        };
        for (i, c) in (0..81).zip(line.chars()) {
            let x = c.to_digit(10).unwrap() as u8;
            if x == 0 {
                continue;
            }
            s.set_value(val_to_field(x), i)
        }
        s.solve();
        count += 1;
        if count % 1000 == 0 {
            println!("  {count}");
        }
    }
}

#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Args {
    #[arg(short, long, default_value = "puzzle.txt")]
    filename: String,
    #[arg(short, long, default_value_t = 1)]
    count: usize,
    #[arg(short, long, action = clap::ArgAction::SetTrue)]
    parallel: bool,
    #[arg(short, long, action = clap::ArgAction::SetTrue)]
    alt_format: bool,
}

fn main() {
    let args = Args::parse();

    let file = Arc::new(args.filename);
    let count = args.count;

    if args.alt_format {
        solve_alt_format(&file);
        return;
    }

    println!("Solving {file} {count} times");


    let s =
    if args.parallel {
        println!("Parallel");
        let thread_count = available_parallelism().unwrap().get();
        println!("Spawning {thread_count} threads");

        // Load initial sudoku once and share it across threads
        let initial_sudoku = Arc::new(load_sudoku(&file));
        let mut handles: Vec<thread::JoinHandle<()>> = vec![];
        let shared_results = Arc::new(Mutex::new(Vec::new()));

        for i in 0..thread_count {
            let thread_results = Arc::clone(&shared_results);
            let thread_initial = Arc::clone(&initial_sudoku);
            let n = count / thread_count + if i < (count % thread_count) {1} else {0};
            let handle = thread::spawn(move || {
                let mut s = thread_initial.as_ref().clone();
                s.solve();
                for _ in 1..n {
                    s = thread_initial.as_ref().clone();
                    s.solve();
                }
                let mut results = thread_results.lock().unwrap();
                results.push(s);
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }
        let final_results = shared_results.lock().unwrap();
        final_results[0]

    } else {
        solve_n(&file, args.count)
    };
    s.print();
}

File: sudoku.zig
const std = @import("std");

// The bit field type used to represent possible values (1-9) in a cell
// Using u16 for consistency with other implementations
const Field = u16;

const max_eager = 91;

// Structure to represent a Sudoku puzzle
const Sudoku = struct {
    // The final derived value for each cell. 0 if still unknown or a bitfield with a single
    // bit set if the solution has been found.
    cells: [81]Field,

    // Remaining possibilities for each block, column, or row.
    blocks: [9]Field,
    cols: [9]Field,
    rows: [9]Field,

    // Number of remaining unsolved cells
    remaining: usize,

    // Initialize the Sudoku with all cells empty and all possibilities available
    pub fn init() Sudoku {
        // Bits 1 through 9 are set since any of those might be a valid guess to start with
        const start_possibilities: Field = 0x1FF;
        return Sudoku{
            .cells = [_]Field{0} ** 81,
            .blocks = [_]Field{start_possibilities} ** 9,
            .cols = [_]Field{start_possibilities} ** 9,
            .rows = [_]Field{start_possibilities} ** 9,
            .remaining = 81,
        };
    }

    // Set a value at a specific location
    pub fn setValue(self: *Sudoku, val: Field, location: usize) void {
        std.debug.assert(self.cells[location] == 0);
        self.cells[location] = val;

        const indices = getIndices(location);
        self.blocks[indices.blk] &= ~val;
        self.cols[indices.col] &= ~val;
        self.rows[indices.row] &= ~val;
        self.remaining -= 1;
    }

    // Clear a value at a specific location
    pub fn clearValue(self: *Sudoku, location: usize) void {
        std.debug.assert(self.cells[location] != 0);
        const val = self.cells[location];
        self.cells[location] = 0;

        const indices = getIndices(location);
        self.blocks[indices.blk] |= val;
        self.cols[indices.col] |= val;
        self.rows[indices.row] |= val;
        self.remaining += 1;
    }

    // Back out moves from a list of locations
    pub fn backOutMoves(self: *Sudoku, eager_moves: [max_eager]usize, eager_count: usize) void {
        for (eager_moves[0..eager_count]) |guess| {
            self.clearValue(guess);
        }
    }

    // Solve the Sudoku puzzle
    pub fn solve(self: *Sudoku) bool {
        // Stage once, do as many forced moves as possible
        // Keep trying until they stop coming
        var eager_moves: [max_eager]usize = undefined;
        var eager_count: usize = 0;

        while (true) {
            var solved_count: usize = 0;
            var lowest_space: usize = 0;
            var lowest_count: u32 = std.math.maxInt(u32);
            var lowest_possibles: Field = 0;

            for (0..81) |i| {
                if (self.cells[i] != 0) {
                    // This cell is already solved
                    continue;
                }
                const indices = getIndices(i);
                const possibles = self.blocks[indices.blk] & self.cols[indices.col] & self.rows[indices.row];
                const count = @popCount(possibles);

                if (count == 1) {
                    self.setValue(possibles, i);
                    solved_count += 1;
                    eager_moves[eager_count] = i;
                    eager_count += 1;
                } else if (count == 0) {
                    // We're down a blind alley, abort
                    self.backOutMoves(eager_moves, eager_count);
                    return false;
                } else if (count < lowest_count) {
                    lowest_count = count;
                    lowest_space = i;
                    lowest_possibles = possibles;
                }
            }

            if (solved_count != 0) {
                // If we're finding moves by elimination don't start guessing yet, just keep on
                // solving this way
                continue;
            } else if (self.remaining == 0) {
                // We won!
                return true;
            }

            const possibles = fieldToVals(lowest_possibles);

            for (possibles.vals[0..possibles.count]) |guess| {
                self.setValue(valToField(guess), lowest_space);
                if (self.solve()) {
                    return true;
                }
                self.clearValue(lowest_space);
            }

            self.backOutMoves(eager_moves, eager_count);
            // None of the guesses worked, we're on a bad branch. Abort
            return false;
        }
    }

    // Print the Sudoku board
    pub fn print(self: *const Sudoku) !void {
        const stdout_file = std.fs.File.stdout();
        var buffer: [2048]u8 = undefined;
        var stdout = stdout_file.writer(&buffer);

        for (0..81) |i| {
            if (i == 0) {
                // First row
            } else if (i % 27 == 0) {
                stdout.interface.print("\n\n", .{}) catch unreachable;
            } else if (i % 9 == 0) {
                stdout.interface.print("\n", .{}) catch unreachable;
            } else if (i % 3 == 0) {
                stdout.interface.print("   ", .{}) catch unreachable;
            } else {
                stdout.interface.print(" ", .{}) catch unreachable;
            }
            stdout.interface.print("{d}", .{fieldToVal(self.cells[i])}) catch unreachable;
        }
        stdout.interface.print("\n\n", .{}) catch unreachable;
        try stdout.interface.flush();
    }
};

// Struct to hold the block, column, and row indices for a given cell
const Indices = struct {
    blk: usize,
    col: usize,
    row: usize,
};

// Convert a position to its block, column, and row indices
fn getIndices(pos: usize) Indices {
    const row = pos / 9;
    const col = pos % 9;
    const blk = (row / 3) * 3 + (col / 3);
    return Indices{ .blk = blk, .col = col, .row = row };
}

// Convert a value (1-9) to its field representation
fn valToField(x: u8) Field {
    return @as(Field, 1) << @intCast(x - 1);
}

// Convert a field to its value (1-9 or 0 if no bits are set)
fn fieldToVal(x: Field) u8 {
    if (x == 0) {
        return 0;
    }
    std.debug.assert(@popCount(x) == 1);
    return @as(u8, @intCast(@ctz(x))) + 1;
}

// Convert a field to a list of possible values
fn fieldToVals(x: Field) struct { vals: [9]u8, count: usize } {
    var vals: [9]u8 = undefined;
    var count: usize = 0;
    if (x == 0) {
        return .{ .vals = vals, .count = 0};
    }

    var val: u8 = 1;
    var field = x;
    while (field > 0) {
        if ((field & 1) == 1) {
            vals[count] = val;
            count += 1;
        }
        val += 1;
        field >>= 1;
    }

    return .{ .vals = vals, .count = count};
}

// Load a Sudoku from a file
fn loadSudoku(filename: []const u8, allocator: std.mem.Allocator) !Sudoku {
    var s = Sudoku.init();

    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    var buf: [1024]u8 = undefined;

    var reader = file.reader(&buf);

    var numbers: std.ArrayList(u8) = .empty;
    defer numbers.deinit(allocator);

    while (try reader.interface.takeDelimiter('\n')) |line| {
        // Skip empty lines
        if (line.len == 0) continue;

        var it = std.mem.tokenizeAny(u8, line, " \t");
        while (it.next()) |token| {
            if (token.len == 0) continue;

            const x = try std.fmt.parseInt(u8, token, 10);
            try numbers.append(allocator, x);

            // Debugging: print the numbers being read
            // std.debug.print("Read number: {d}\n", .{x});
        }
    }

    if (numbers.items.len < 81) {
        std.debug.print("Not enough numbers in the input file: got {d}, need 81\n", .{numbers.items.len});
        return error.InvalidPuzzle;
    }

    for (numbers.items[0..81], 0..) |x, i| {
        if (x == 0) {
            continue;
        }
        s.setValue(valToField(x), i);
    }

    return s;
}

// Solve a Sudoku puzzle n times and return the last solved puzzle
fn solveN(filename: []const u8, n: usize, allocator: std.mem.Allocator) !Sudoku {
    const original = try loadSudoku(filename, allocator);
    var s = original;

    for (0..n) |_| {
        // Take a copy of the original
        s = original;

        _ = s.solve();
    }

    return s;
}

pub fn main() !void {
    // Set up command line argument parsing
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    defer _ = gpa.deinit();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    var filename: []const u8 = "puzzle.txt";
    var iterations: usize = 1;

    if (args.len > 1) {
        filename = args[1];

        if (args.len > 2) {
            iterations = try std.fmt.parseInt(usize, args[2], 10);
        }
    }

    std.debug.print("Solving {s} {d} times\n", .{ filename, iterations });

    var s = try solveN(filename, iterations, allocator);
    try s.print();
}
